crates/bevy_app/src/app.rs

```diff
@@ -652,6 +652,32 @@ impl App {
         self
     }
 
+    /// Inserts a resource to the current [App] and not overwrites any resource previously added of the same type.
+    ///
+    /// A resource in Bevy represents globally unique data. Resources must be added to Bevy Apps
+    /// before using them. This happens with [`insert_resource`](Self::insert_resource).
+    ///
+    /// See also `init_resource` for resources that implement `Default` or [`FromWorld`].
+    ///
+    /// ## Example
+    /// ```
+    /// # use bevy_app::prelude::*;
+    /// #
+    /// struct MyCounter {
+    ///     counter: usize,
+    /// }
+    ///
+    /// App::new()
+    ///    .insert_resource_not_overwrite(MyCounter { counter: 0 });
+    /// ```
+    pub fn insert_resource_not_overwrite<R: Resource>(&mut self, resource: R) -> &mut Self {
```

Comment:
I wonder if this method name could be shortened. Maybe `try_insert_resource` (although that might imply that it returns an `Option` or something...)?

Reply:
This could mirror [HashMap::try_insert](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.try_insert) and return a `Result`; you may want to know if the insertion was successful or not.

Reply:
> you may want to know if the insertion was successful or not.

That's a good point. The only thing about that is then we can't return `self` for chaining, which isn't ideal.

Reply:
ah, yeah of course. That's a problem

Reply:
> I was actually wanting this functionality just today haha. A few thoughts:
> 
>     * This might be nice to have on `World` itself (to avoid using things like `world.get_resource_or_insert_with(...)`)
> 
>     * It might also be nice to have a similar one for `init_resource` (possibly for the `non_send` ones too?)

non_send should of course also be necessary.

Reply:
If I want to add these changes together. What should I do? This is my first commit.